server/src/main/scala/com/cloudera/livy/server/batch/BatchSession.scala

```diff
@@ -23,62 +23,129 @@ import java.lang.ProcessBuilder.Redirect
-    require(request.file != null, "File is required.")
-
-    val builder = new SparkProcessBuilder(livyConf)
-    builder.conf(conf)
-
-    proxyUser.foreach(builder.proxyUser)
-    request.className.foreach(builder.className)
-    request.driverMemory.foreach(builder.driverMemory)
-    request.driverCores.foreach(builder.driverCores)
-    request.executorMemory.foreach(builder.executorMemory)
-    request.executorCores.foreach(builder.executorCores)
-    request.numExecutors.foreach(builder.numExecutors)
-    request.queue.foreach(builder.queue)
-    request.name.foreach(builder.name)
-
-    // Spark 1.x does not support specifying deploy mode in conf and needs special handling.
-    livyConf.sparkDeployMode().foreach(builder.deployMode)
-
-    builder.redirectOutput(Redirect.PIPE)
-    builder.redirectErrorStream(true)
-
-    val file = resolveURIs(Seq(request.file))(0)
-    val sparkSubmitProcess = builder.start(Some(file), request.args)
-    SparkApp.create(uniqueAppTag, Option(sparkSubmitProcess), livyConf, Some(this))
+    val (appId, sparkSubmit): (Option[String], Option[LineBufferedProcess]) = {
+      def create(request: CreateBatchRequest) = {
+        val conf = SparkApp.prepareSparkConf(
+          appTag,
+          livyConf,
+          prepareConf(request.conf, request.jars, request.files, request.archives, request.pyFiles))
+        require(request.file != null, "File is required.")
+
+        val builder = new SparkProcessBuilder(livyConf)
+        builder.conf(conf)
+
+        proxyUser.foreach(builder.proxyUser)
+        request.className.foreach(builder.className)
+        request.driverMemory.foreach(builder.driverMemory)
+        request.driverCores.foreach(builder.driverCores)
+        request.executorMemory.foreach(builder.executorMemory)
+        request.executorCores.foreach(builder.executorCores)
+        request.numExecutors.foreach(builder.numExecutors)
+        request.queue.foreach(builder.queue)
+        request.name.foreach(builder.name)
+
+        // Spark 1.x does not support specifying deploy mode in conf and needs special handling.
+        livyConf.sparkDeployMode().foreach(builder.deployMode)
+
+        sessionStore.save(BatchSession.RECOVERY_SESSION_TYPE, recoveryMetadata)
+
+        builder.redirectOutput(Redirect.PIPE)
+        builder.redirectErrorStream(true)
+
+        val file = resolveURIs(Seq(request.file))(0)
+        val sparkSubmit = builder.start(Some(file), request.args)
+
+        (None, Option(sparkSubmit))
+      }
+
+      def recover(appId: Option[String]) = (appId, None)
+
+      createOrRecover.fold(create, recover)
+    }
+
+    SparkApp.create(appTag, appId, sparkSubmit, livyConf, Some(this))
   }
 
   protected implicit def executor: ExecutionContextExecutor = ExecutionContext.global
 
-  private[this] var _state: SessionState = SessionState.Starting()
+  private[this] var _state: SessionState =
+    createOrRecover.fold(_ => SessionState.Starting(), _ => SessionState.Recovering())
```

Comment:
Hmm... if you follow my suggestion this would become a constructor parameter... (`initialState` or something.) Still I think it's worth it.